### PR TITLE
aio: add monitoring and fix a couple of issues

### LINKED ISSUE #21942: aio: add regular e2e testing of angular.io
https://support.circleci.com/hc/en-us/articles/115015481128-Scheduling-jobs-cron-for-builds-

### PR DESCRIPTION
**Pull request by gkalpak**
- **Fix: Wait for the app to stabilize before registering the SW.**
   Wait for the app to stabilize, before registering the SW and starting to check for SW updates. This avoids setting up a long timeout, which would prevent the app from stabilizing and thus cause issues with Protractor.

- **Fix: Fix SW routing RegExp to allow redirecting `/api/animate` URLs.**
    Prevent the SW from routing URLs starting with `/api/animate` to `index.html`. Instead, pass them through to the server, where they are correctly redirected.

- **CI: Add monitoring for angular.io.**
    Configure a periodic job to be run on CircleCI, performing several checks against the actual apps deployed to production (https://angular.io) and staging (https://next.angular.io).

Fixes #21942.

### CODE REVIEW

File: aio/src/main.ts [hunk 1]
```diff
@@ -11,7 +13,7 @@ if (environment.production) {
 platformBrowserDynamic().bootstrapModule(AppModule).then(ref => {
   if (environment.production && 'serviceWorker' in (navigator as any)) {
     const appRef: ApplicationRef = ref.injector.get(ApplicationRef);
-    appRef.isStable.first().subscribe(() => {
+    appRef.isStable.filter(v => v).first().subscribe(() => {
```
Comment (petebacondarwin) [inline aio/src/main.ts]: Can be simplified to:

```ts
appRef.isStable.first(v => v).subscribe(() => {
```

File: aio/tests/deployment-config/unit/testFirebaseRedirection.spec.ts [hunk 2]
```diff
@@ -1,8 +1,8 @@
-import { getRedirector, loadLegacyUrls, loadRedirects, loadSitemapUrls } from './helpers';
+import { getRedirector, loadLegacyUrls, loadRedirects, loadLocalSitemapUrls } from '../shared/helpers';
```
Comment (petebacondarwin) [inline aio/tests/deployment-config/unit/testFirebaseRedirection.spec.ts]: AFAICT `loadLocalSitemapUrls` does not exist...

File: aio/tests/deployment-config/shared/URLS_TO_REDIRECT.txt [hunk 3]
```diff
@@ -180,3 +182,8 @@
 /news.html	https://blog.angular.io/
 /testing	/guide/testing
 /testing/first-app-tests.html	/guide/testing
+
+# These URLs are deliberately put at the end in order to ensure that the ServiceWorker will have
+# been installed. This allows verifying that the SW will indeed let them pass through to the server.
```
Comment (petebacondarwin) [inline aio/tests/deployment-config/shared/URLS_TO_REDIRECT.txt]: Are you relying upon the SW loading during processing of these tests? Why else would it matter that they are at the end of the file?

File: aio/package.json [hunk 4]
```diff
@@ -57,12 +58,11 @@
     "generate-zips": "node ./tools/example-zipper/generateZips",
     "sw-manifest": "ngu-sw-manifest --dist dist --in ngsw-manifest.json --out dist/ngsw-manifest.json",
     "sw-copy": "cp node_modules/@angular/service-worker/bundles/worker-basic.min.js dist/",
-    "postinstall": "node tools/cli-patches/patch.js && uglifyjs node_modules/lunr/lunr.js -c -m -o src/assets/js/lunr.min.js --source-map",
     "build-ie-polyfills": "node node_modules/webpack/bin/webpack.js -p src/ie-polyfills.js src/generated/ie-polyfills.min.js",
+    "update-webdriver": "webdriver-manager update --standalone false --gecko false $CHROMEDRIVER_VERSION_ARG",
```
Comment (petebacondarwin) [inline aio/package.json]: AFAICT renaming `~update-webdriver` to `update-webdriver` is the only real change in this file, right? Or am I missing something?

File: aio/tests/deployment-config/shared/helpers.ts [hunk 5]
```diff
@@ -34,6 +28,27 @@ export function loadLegacyUrls() {
       .map(line => line.split('\t'));            // Split lines into URL pairs.
 }
 
+export function loadLocalSitemapUrls() {
```
Comment (petebacondarwin) [inline aio/tests/deployment-config/shared/helpers.ts]: Ahah! I see that the non-existent `loadLocalSitemapUrls` was a rebase mismatch.

**Review by petebacondarwin — APPROVED** [review]
A few minor nits that you can choose to ignore but otherwise looks good.
I like what you did with redirect and sitemap testing !

File: aio/tests/deployment-config/unit/testFirebaseRedirection.spec.ts (hunk 2 above)
Comment (petebacondarwin) [reply aio/tests/deployment-config/unit/testFirebaseRedirection.spec.ts]: I see it was added in a later commit :-0

Comment (gkalpak) [reply aio/tests/deployment-config/unit/testFirebaseRedirection.spec.ts]: Oops :grin:

File: aio/tests/deployment-config/shared/URLS_TO_REDIRECT.txt (hunk 3 above)
Comment (gkalpak) [reply aio/tests/deployment-config/shared/URLS_TO_REDIRECT.txt]: Yes, I rely on the SW being loaded.

Essentially, if you put these URLs at the top of the file (where they belong alphabetically :grin:), they are navigatd to before the SW has been installed. As a result, they pass through to the server, where they are redirected correctly.

Once the SW has been activated and handles navigation, these URLs will only pass through to the server if they are not matched by the routing RegExp in `ngsw-manifest.json`.

##
In other words:
Having the URLs at the top would allow the tests to pass even with the current (broken) RegExp on master.
Having the URLs at the end, makes the tests fail on master (and pass with the fix in this PR).

Comment (petebacondarwin) [reply aio/tests/deployment-config/shared/URLS_TO_REDIRECT.txt]: What? Surely not?
Is there not some other mechanism to ensure that the SW has been activated before running these tests?
Perhaps a dummy request in the protractor spec or something?

Comment (gkalpak) [reply aio/tests/deployment-config/shared/URLS_TO_REDIRECT.txt]: I just realized my comment might be confusing, because I am referring to the e2e tests added in the _next_ commit :grin:
The jasmine tests (run with `yarn deployment-config-test`) will correctly fail/pass regardless where you put the URLs :grin:

File: aio/package.json (hunk 4 above)
Comment (gkalpak) [reply aio/package.json]: Pretty much 😇

File: aio/src/app/sw-updates/sw-updates.service.ts [hunk 6]
```diff
@@ -37,10 +37,12 @@ export class SwUpdatesService implements OnDestroy {
       .filter(({type}) => type === 'activation')
       .map(({version}) => version);
 
-  constructor(private logger: Logger, private sw: NgServiceWorker) {
-    this.checkForUpdateSubj
-        .debounceTime(this.checkInterval)
-        .startWith(null)
+  constructor(appRef: ApplicationRef, private logger: Logger, private sw: NgServiceWorker) {
+    const appIsStable$ = appRef.isStable.filter(v => v).first();
```
[the remark this answers is not in the document]
Comment (gkalpak) [reply aio/src/app/sw-updates/sw-updates.service.ts]: I had no idea about this 💯

File: aio/tests/deployment-config/shared/URLS_TO_REDIRECT.txt (hunk 3 above)
Comment (petebacondarwin) [reply aio/tests/deployment-config/shared/URLS_TO_REDIRECT.txt]: I understand but I still think it is rather flakey to rely upon the order of the test data in this way.

File: aio/src/app/sw-updates/sw-updates.service.ts [hunk 7]
```diff
@@ -60,7 +62,7 @@ export class SwUpdatesService implements OnDestroy {
     this.sw.checkForUpdate()
         // Temp workaround for https://github.com/angular/mobile-toolkit/pull/137.
         // TODO (gkalpak): Remove once #137 is fixed.
-        .concat(Observable.of(false)).take(1)
+        .concat(of(false)).first()
```
[the remark this answers is not in the document]
Comment (gkalpak) [reply aio/src/app/sw-updates/sw-updates.service.ts]: I didn't know about this one either. Nice!
(I see someone knows their RxJS :smiley:)

File: aio/tests/deployment-config/shared/URLS_TO_REDIRECT.txt (hunk 3 above)
Comment (gkalpak) [reply aio/tests/deployment-config/shared/URLS_TO_REDIRECT.txt]: I know 😞
(TBH, I don't think it matters, because we already assert that these URLs are handled correctly by the SW in `depoyment-config-test`, but I will look into it.)

Comment (gkalpak) [reply aio/tests/deployment-config/shared/URLS_TO_REDIRECT.txt]: I didn't find a reliable way to ensure that the ServiceWorker has been activated, but I also think that we mainly want to check the Firebase server (since we are pretty confident about the SW routing RegExp based on the tests in `deployment-config-test`).

So, I changed the code to instead unregister the SW [before each test][1] :grin:

[1]: https://github.com/angular/angular/pull/22483/files#diff-f4ab6430117a395c91e573a3e676cb0dR10

**Comment by gkalpak** [comment]
@petebacondarwin, I added a bunch of fixup commits. (I didn't touch the non-fixup commits other than rebasing on master.) PTAL

**Review by petebacondarwin — APPROVED** [review]

File: aio/tests/deployment-config/e2e/testDeployedRedirection.e2e-spec.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+  describe('(with sitemap URLs)', () => {
+    sitemapUrls.forEach((url, i) => {
+      it(`should not redirect '${url}' (${i + 1}/${sitemapUrls.length})`, async () => {
+        browser.get(url);
+
+        const expectedUrl = browser.baseUrl + url;
+        const actualUrl = (await browser.getCurrentUrl()).replace(/\?.*$/, '');
+
+        expect(actualUrl).toBe(expectedUrl);
```
Comment (IgorMinar) [inline aio/tests/deployment-config/e2e/testDeployedRedirection.e2e-spec.ts]: shouldn't we also verify that we didn't render an error or 404 page instead?

if you could add an assertion for the content of the doc-viewer DOM element I'd trust this test more.

**Review by IgorMinar — APPROVED** [review]
this looks great otherwise. I just wonder if testing all the urls is not going to take too long and be too flaky. Have you considered that?

File: aio/tests/deployment-config/e2e/testDeployedRedirection.e2e-spec.ts (hunk 8 above)
Comment (gkalpak) [reply aio/tests/deployment-config/e2e/testDeployedRedirection.e2e-spec.ts]: Actually, while this would be nice, it would require waiting for the app to stabilize, which in turn would dramatically increase the total time (~5x).

**Comment by gkalpak** [comment]
Pushed another fixup commit.

##
Long running times are indeed a concern. After I realized I was doing an extra request per sitemap/legacy URL, the time for those tests was brought down to ~7.5mins (from ~17mins), which brings the overall time down to ~11mins per app URL. So, it would be ~22mins (plus setup time) for our two URLs (angular.io and next.angular.io), which sounds reasonable for this kind of job.

If we want to reduce these times further, we could consider probing some of the URLs only. (But I would avoid randomness to retain reproducibility. Maybe we can test a subset of the URLs based on the current date or something.)

##
I've not seen any flakeyness, when running this locally, so I think it is reasonable to not expect any more flakeyness that normally have on CI jobs :grin:
But, only time will tell for sure 🤔

**Comment by petebacondarwin** [comment]
If this is a CRON job that doesn't run when we are pushing commits and PRs, does it matter that it takes a long time? (Other than using up resources once a day for a bit)

**Comment by gkalpak** [comment]
Yes, this will only run once per day. My main concern is that the job doesn't exceed the max allowed running time. Not sure what that is on CircleCI, but on Travis for example jobs can't exceed 50mins. So, you can't have a job running for 2 hours :grin:

If we are absolutely want to do the slow checks on all URLs, we can break it down to multiple jobs (as we do for the aio examples e2e tests).

**Comment by petebacondarwin** [comment]
Also, there is nothing stopping us from running each domain in a different job (e.g. angular.io in one and next.angular.io in a different job).

**Comment by gkalpak** [comment]
Thx, @ngbot! Any chance you can help out by resolving the conflicts yourself?

File: aio/tests/deployment-config/e2e/testDeployedRedirection.e2e-spec.ts (hunk 8 above)
Comment (IgorMinar) [reply aio/tests/deployment-config/e2e/testDeployedRedirection.e2e-spec.ts]: In that case we should add these more expensive assertions just for a smaller sample of urls. maybe in a separate pr?

**Review by IgorMinar — APPROVED** [review]

File: aio/tests/deployment-config/e2e/testDeployedRedirection.e2e-spec.ts (hunk 8 above)
Comment (gkalpak) [reply aio/tests/deployment-config/e2e/testDeployedRedirection.e2e-spec.ts]: I created #23063 to track that. Let's land this and see how it works for a few days first.

**Comment by alxhub** [comment]
This doesn't apply cleanly to the patch branch (conflicts with Rx pipeable operators update). Please split it into two PRs :) Thanks!

**Comment by gkalpak** [comment]
OK. Let's merge this into master and I'll submit another PR for 5.2.x.
